### PR TITLE
Refactor DbConnectionValidator

### PR DESCRIPTION
**Pull request by hishamco**

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs [hunk 1]
```diff
@@ -125,63 +156,29 @@ public async Task<DbConnectionValidatorResult> ValidateAsync(string databaseProv
...
-    private ISqlBuilder GetSelectBuilderForShellDescriptorDocument(string tablePrefix, DatabaseProviderName providerName)
-    {
-        var selectBuilder = GetSqlBuilder(providerName, tablePrefix);
-
-        selectBuilder.Select();
-        selectBuilder.Selector("*");
-        selectBuilder.Table(_tableNameConvention.GetDocumentTable());
-        selectBuilder.WhereAnd($"Type = '{_shellDescriptorTypeColumnValue}'");
-        selectBuilder.Take("1");
+        static ISqlBuilder GetDocumentTableSelectBuilder(string tablePrefix, string tablePrefixSeparator, string documentTable, ISqlDialect dialect, bool isShellDescriptorDocument = false)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:160]: Can you move this outside of the `ValidateAsync` method and use the `private` keyword?

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs [hunk 2]
```diff
@@ -63,7 +63,30 @@ public async Task<DbConnectionValidatorResult> ValidateAsync(string databaseProv
             return DbConnectionValidatorResult.InvalidConnection;
         }
 
-        var factory = GetFactory(providerName, connectionString);
+        IConnectionFactory factory = null;
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:66]: Can you extract this into a method just so the `ValidateAsync` isn't a 100 lines long.

**Review by MikeAlhayek — COMMENTED** [review]
I personally don't like how this method is super long. Few suggestions to help with that

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs (hunk 1 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:160]: Seems `GetDocumentTableSelectBuilder()` make it a little bit bigger, I will extract it

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs [hunk 3]
```diff
@@ -63,7 +63,29 @@ public async Task<DbConnectionValidatorResult> ValidateAsync(string databaseProv
             return DbConnectionValidatorResult.InvalidConnection;
         }
 
-        var factory = GetFactory(providerName, connectionString);
+        var (factory, dialect) = GetConnectionFactoryAndSqlDialect(providerName, connectionString);
+
+        switch (providerName)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:68]: This can be removed now since you created `GetConnectionFactoryAndSqlDialect`

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs]: Seems I'm out of coverage at that time ;)

**Comment by MikeAlhayek** [comment]
@hishamco I am approving this. However, there is not huge added value here. More about refactoring the DbConnectionValidator in  #12657
If this isn't merged soon, we'll see conflicts as @jtkech  and I discussing ways to actually improve this class not just simply refactoring the code.

**Review by MikeAlhayek — APPROVED** [review]

**Comment by hyzx86** [comment]
Can this PR fix this validation?

image
Server=192.168.0.200;Database=xxxx;uid=xxx;Password=xxxx

**Comment by hishamco** [comment]
@hyzx86 this PR is not related to the above issue, could you please file a new one? @MikeAlhayek please handle this while you worked on this

**Comment by hyzx86** [comment]
Yes, it may be related to the update of Yessql
We need to add Encryption=false or TrustServerCertificate=True to the connection string

**Comment by hishamco** [comment]
I pointed to that [here](https://github.com/OrchardCMS/OrchardCore/issues/12679#issuecomment-1287366934), FYI @Skrypt already did a PR for adding docs for this one, please check the PR #12681

**Comment by hishamco** [comment]
@jtkech your approval if you don't have any comments

**Comment by jtkech** [comment]
Oops sorry, many change around this in #12683 on which I'm working on.

**Comment by Skrypt** [comment]
@jtkech please close this one or advise.

**Comment by jtkech** [comment]
@Skrypt 

Hmm because #12683 is now merged, @hishamco could merge dev, solve the conflicts, and then redo the same kind of refactoring, but as he wants, otherwise yes we will close it.

**Comment by hishamco** [comment]
@jtkech it's ready to merge

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs [hunk 4]
```diff
@@ -164,50 +163,32 @@ public async Task<DbConnectionValidatorResult> ValidateAsync(DbConnectionValidat
...
-            _ => throw new ArgumentOutOfRangeException(nameof(databaseProvider), "Unsupported database provider"),
-        };
-    }
-
-    private static ISqlDialect GetSqlDialect(string databaseProvider)
-    {
-        return databaseProvider switch
+    private static (IConnectionFactory connectionFactory, ISqlDialect sqlDialect) GetFactoryAndSqlDialect(
+        string databaseProvider,
+        string connectionString) => databaseProvider switch
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:184]: I'd agree since I prefer expression-bodied members, but in OC we deliberately don't use them (and it's a bit easier to read in this case too).

**Review by Piedone — CHANGES_REQUESTED** [review]

**Review by Piedone — APPROVED** [review]
Merge conflict needs to be solved but otherwise LGTM.

**Comment by hyzx86** [comment]
Do we need to set `MultipleActiveResultSets=True;`  add to default configuration?

`Putting the workflow in the faulted state. System.InvalidOperationException: The connection does not support MultipleActiveResultSets`

**Comment by hishamco** [comment]
@MikeAlhayek anything to add or shall we merge this?

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs [hunk 5]
```diff
@@ -161,14 +160,18 @@ public async Task<DbConnectionValidatorResult> ValidateAsync(DbConnectionValidat
...
-    private static SqlBuilder GetSelectBuilderForDocumentTable(SqlBuilder sqlBuilder, string documentTable, string schema)
+    private static string GetDocumentCommandText(ISqlBuilder sqlBuilder, string documentTable, string schema, bool isShellDescriptorDocument = false)
     {
         sqlBuilder.Select();
         sqlBuilder.Selector("*");
         sqlBuilder.Table(documentTable, alias: null, schema);
-        sqlBuilder.Take("1");
 
-        return sqlBuilder;
+        if (isShellDescriptorDocument)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:169]: @hishamco why is this needed? Also why did you remove Take(1) out? Take(1) to improve the query speed since the server will stop looking after it finds the first record.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:175]: It's needed to check the type value

https://github.com/OrchardCMS/OrchardCore/blob/71fde7c22a098d2003b196b2ad013516e7c1af98/src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs#L185

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:175]: > Also why did you remove Take(1) out?

Oops

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:175]: I already did

**Comment by hishamco** [comment]
@MikeAlhayek is there anything else you need to add or shall I merge this?

**Comment by MikeAlhayek** [comment]
@hishamco  move `sqlBuilder.Take("1");` to line 174 please and merge it.
